amethyst_network/src/network_socket.rs

```diff
@@ -69,89 +50,91 @@ where
+        if config.udp_recv_addr.port() < 1024 {
             // Just warning the user here, just in case they want to use the root port.
             warn!("Using a port below 1024, this will require root permission and should not be done.");
         }
 
-        let mut socket = UdpSocket::bind(addr, NetworkConfig::default())
-            .map_err(|x| Error::new(ErrorKind::Other, x.to_string()))?;
+        let server = Host::run(&config)?;
 
-        socket.set_nonblocking(true).map_err(|_| {
-            Error::new(
-                ErrorKind::Other,
-                "Unable to set `UdpSocket` to non-blocking mode",
-            )
-        })?;
+        let udp_send_handle = server.udp_send_handle();
+        let udp_receive_handle = server.udp_receive_handle();
 
-        // this -> thread
-        let (tx1, rx1) = channel();
-        // thread -> this
-        let (tx2, rx2) = channel();
+        let server_sender = NetSocketSystem::<E>::start_sending(udp_receive_handle);
+        let server_receiver = NetSocketSystem::<E>::start_receiving(udp_receive_handle);
 
-        thread::spawn(move || {
-            //rx1,tx2
-            let send_queue = rx1;
-            let receive_queue = tx2;
-            let mut socket = socket;
-
-            'outer: loop {
-                // send
-                for control_event in send_queue.try_iter() {
-                    match control_event {
-                        InternalSocketEvent::SendEvents { target, events } => {
-                            for ev in events {
-                                send_event(&ev, &target, &mut socket);
-                            }
+        Ok(NetSocketSystem {
+            filters,
+            transport_sender: server_sender,
+            transport_receiver: server_receiver,
+            config,
+        })
+    }
+
+    /// Start a thread to send all queued packets.
+    fn start_sending(sender: Arc<SendHandler>) -> Sender<InternalSocketEvent<E>> {
+        let (tx, send_queue) = mpsc::channel();
+
+        thread::spawn(move || loop {
+            for control_event in send_queue.try_iter() {
+                match control_event {
+                    InternalSocketEvent::SendEvents { target, events } => {
+                        for ev in events {
+                            send_event(ev, target, &sender.get_sender());
                         }
-                        InternalSocketEvent::Stop => break 'outer,
+                    }
+                    InternalSocketEvent::Stop => {
+                        break;
                     }
                 }
+            }
+        });
+
+        tx
+    }
 
-                // receive
+    /// Starts a thread which receives incoming packets and sends them onto the 'Receiver' channel.
+    fn start_receiving(receiver: Arc<Mutex<ReceiveHandler>>) -> Receiver<Packet> {
+        let (receive_queue, rx) = mpsc::channel();
+
+        thread::spawn(move || {
+            if let Ok(receiver) = receiver.lock() {
```

Comment:
this lock will never be unlocked unless there is a send error, this seems a bit odd? Though I think this function will only get called once? So I think we should either document that or panic if it gets called twice.

Reply:
Indeed, this will not be possible. 

The whole point why the `Receiver` is not sync is that there could not be different receivers. This is the case here. We have one receiving thread, and so requiring the lock for the receiver there is a valid use case. If someone else tries to receive it is not possible. And if he does he should not be able.

Reply:
Yeah what I’m saying is that we should be explicit about this behavior so there should be a comment about it and a panic if it gets called more than once.

Reply:
Yes, I'll add that